Number formatting using Intl.* APIs

niik opened the pull request:
<!--
What GitHub Desktop issue does this PR address? (for example, #1234)
-->

Closes #1245

## Description
<!--
A summary of the changes made along with any other information that would be helpful to a reviewer such as potential tradeoffs or alternative approaches you considered.
-->

~~Back in 2017 I wrote *"I'm leaning towards trying to solve this properly."* in response to our first attempt  at fixing #1245 (https://github.com/desktop/desktop/pull/1257). Now, 5-ish years later technology has finally caught up to where we're able to do this with a relatively small amount of support logic :tada:~~

**Turns out ☝️ was a little bit premature, see https://github.com/desktop/desktop/pull/14173#issuecomment-1072293321**

This builds upon #14173 and adds region-aware formatting of numbers.

I also took this opportunity to render the number of commits ahead/behind in a compact format (i.e number > 1000 will render as 1K). Hovering will get you the exact number but if you're behind by 1000 or more you likely don't care about the exact number.

There's a good chance I've missed a number of (pun intended) places where we present numbers to the user but that's okay, we can fix that incrementally.

### Screenshots

<!--
If this PR touches the UI layer of the app, please include screenshots or animated gifs to show the changes.
-->

| Before | en-US | en-SE |
|--------|-------|-------|
|  <img width="239" alt="image" src="https://user-images.githubusercontent.com/634063/158374912-6c26db3b-c0f0-4e7e-9fc7-3a5b6815804c.png"> | <img width="248" alt="image" src="https://user-images.githubusercontent.com/634063/158375549-41b264e1-78a9-4d4a-bf18-238b6ef3d7da.png"> | <img width="244" alt="image" src="https://user-images.githubusercontent.com/634063/158374976-0a717bad-276f-43aa-8c13-e627fd206264.png"> |
| <img width="211" alt="image" src="https://user-images.githubusercontent.com/634063/158375173-449f0cdb-96b0-4b1b-a56b-28750c03e75b.png"> | <img width="216" alt="image" src="https://user-images.githubusercontent.com/634063/158375616-eaba13c0-149a-4457-8e0e-604d054af3ad.png"> | <img width="217" alt="image" src="https://user-images.githubusercontent.com/634063/158375250-72be325d-ce7a-4d93-b04a-6f2fee1e10fd.png"> |

## Release notes

<!--
You can leave this blank if you're not sure.
If you don't believe this PR needs to be mentioned in the release notes, write "Notes: no-notes".
-->

Notes: no-notes

tidy-dev commented on app/src/lib/format-number.ts:
```diff
@@ -0,0 +1,37 @@
+import QuickLRU from 'quick-lru'
+import { getFormattingLocales } from './formatting-locale'
+
+// Initializing a nuumber formatter is expensive but formatting is relatively
```
```suggestion
// Initializing a number formatter is expensive but formatting is relatively
```

tidy-dev reviewed (changes requested):
Just a little spelling nit. But, also, a place missed is line 654 of `compare.tsx` for the squashing dialog. 

As for it working, I switched my region to Sweden, Spain, and France. It worked as expected for Sweden, but not for Spain or France. But, I figure the is the same issue that Sergio mentioned in slack so I am assuming not something for this PR.

niik commented:
So this will also be affected by https://github.com/desktop/desktop/pull/14173#issuecomment-1072293321.

Given that we're switching to `en-US-POSIX` by default there will be no number grouping for now.

@sergiou87 @tidy-dev One thing that will change with this PR though is the compact format of commits > 1000. I think that's okay though even though that'll use `.` as the decimal separator. Our plan is definitely to try and support non-US region formats soon and the number of users pushing/pulling more than 1000 commits should be relatively low.

tidy-dev reviewed (approved):
:sparkles: Makes sense. :)

angusdev committed:
Issue 15284 Toggle Diff Display Mode and Hide Whitespace Changes

Eminealtug commented:
.

niik committed:
Set aria-multiselectable when more than one item is selectable

niik committed:
Only set aria-selected for selectable items

See https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-selected

sergiou87 committed:
Remove unnecessary cast to any

sergiou87 committed:
Set `aria-haspopup` prop

sergiou87 committed:
(WIP) Some changes to tag autocomplete items properly

sergiou87 committed:
Make screen readers read autocomplete suggestions

sergiou87 committed:
Fix unknown user autocomplete item

sergiou87 committed:
Fix a11y attributes for autocomplete items

sergiou87 committed:
Make screen reader read number of co-author suggestions

sergiou87 committed:
Some cleanup (or not)

sergiou87 committed:
(WIP) Try things

sergiou87 committed:
First steps to make AutocompletingTextInput accessible

sergiou87 committed:
(WIP) Use AutocompletingInput for co-authors

sergiou87 committed:
Some work to make the Co-Authors textbox look good

sergiou87 committed:
Use autocompletion to get co-authors!

sergiou87 committed:
Create Co-Author autocompletion provider

sergiou87 committed:
Add support for "unknown" users in our autocompletion component

sergiou87 committed:
Only show "Search user" entry when needed

sergiou87 committed:
Always show autocomplete for co-authors

sergiou87 committed:
Use different IDs when autocomplete list is refreshed

sergiou87 committed:
Add aria-live with number of suggestions

sergiou87 committed:
More co-author input improvements. Needs cleanup!

sergiou87 committed:
Add name to added handle aria label

tidy-dev committed:
Merge pull request #16274 from desktop/commit-attribution-descriptive-links

Commit attribution descriptive links

tidy-dev committed:
Add tabIndex -1 to menu items

tidy-dev committed:
Make button be a menuitem

sergiou87 committed:
Bubble up unknown authors instead of keeping them as internal state

tidy-dev committed:
On menu open through space/enter, select first item

tidy-dev committed:
Stop pane from stealing focus from first selected item

sergiou87 committed:
Add icons for unknown authors added

sergiou87 committed:
Search for unknown users to add them as co-authors!

sergiou87 committed:
Clean up

sergiou87 committed:
Open autocomplete list without selected item

sergiou87 committed:
Take username casing into account

sergiou87 committed:
Detect use of space at the end of input and attempt autocomplete

sergiou87 committed:
More cleanup

sergiou87 committed:
Improve types

sergiou87 committed:
More and better comments

sergiou87 committed:
Display autocomplete above element when needed

sergiou87 committed:
Select last author instead of deleting it when pressing backspace

sergiou87 committed:
Be smart about next selected author after removing one

tidy-dev committed:
Add overflow-y to dialogs

tidy-dev committed:
Make preferences dialog scroll vertically when zoomed

niik committed:
Move uncommitted changes preference to prompts tab

dependabot[bot] committed:
Bump webpack from 5.68.0 to 5.76.0

Bumps [webpack](https://github.com/webpack/webpack) from 5.68.0 to 5.76.0.
- [Release notes](https://github.com/webpack/webpack/releases)
- [Commits](https://github.com/webpack/webpack/compare/v5.68.0...v5.76.0)

---
updated-dependencies:
- dependency-name: webpack
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>

sergiou87 committed:
Improve accessibility of AutocompletingTextInput component

sergiou87 committed:
Remove : from emoji suggestion list

sergiou87 committed:
Fix issue autocompletion style and accessibility

niik committed:
Get rid of broken mouse over event in lists

The `focusOnHover` only comes into play when `selectOnHover` is set and the only component using that (now) is the `AutoCompletingTextInput`.

Due to a bug in the mouseOver event handler on the List component we would only invoke the `selectionChanged` event and not the `selectedRowChanged` event meaning that hovering wouldn't ever change the selection. The only thing it would do is scroll to partially visible rows which isn't a common behavior and not worth keeping all this logic (and a11y lint exception) around for.

niik committed:
Allow passing arbitrary a11y props to list rows

tidy-dev committed:
Add aria-haspopup to menu bar buttons

tidy-dev committed:
Docs to aria attributtes

niik committed:
Merge pull request #16313 from desktop/preference-tweaks

Preference dialog, now 99% less jarring

tidy-dev committed:
Add aria-labellby to windows menus

tidy-dev committed:
Merge pull request #16317 from desktop/scrollable-dialogs

Make Preferences Dialog Scroll vertically when zoomed

tidy-dev committed:
Remove focus logic - I think is unneeded.

tidy-dev committed:
Add menu items to resize main view width

tidy-dev committed:
Add the functionality to app store

tidy-dev committed:
Docs..

tidy-dev committed:
lowercase w

sergiou87 committed:
Merge pull request #16324 from desktop/autocompleting-input-a11y

Improve accessibility of `AutocompletingTextInput` component

sergiou87 committed:
Merge branch 'development' into co-author-accessibility-fixes

sergiou87 committed:
Fix typo

sergiou87 committed:
Add new stats for PR comment notifications

sergiou87 committed:
Record new PR comment notification stats

sergiou87 committed:
Disable select on hover for autocompletion lists

sergiou87 committed:
Use author username/name as handle `key`

sergiou87 committed:
Fix check when updating autocompleting list :facepalm:

sergiou87 committed:
Switch focus after component is rendered

sergiou87 committed:
Announce when authors are added/removed

sergiou87 committed:
Co-authors labelled with all added co-authors

sergiou87 committed:
Fix author labelling for screen reader

Co-Authored-By: Pablo Núñez <pablonete@github.com>

sergiou87 committed:
Move code around

sergiou87 committed:
Fix labelling of user not found

JosephTLyons committed:
Add Zed as an external editor option

steveward committed:
Add new RStudio bundle identifier

sergiou87 committed:
Merge pull request #16320 from desktop/dependabot/npm_and_yarn/webpack-5.76.0

Bump webpack from 5.68.0 to 5.76.0

tidy-dev committed:
Announce Committing Status/Move to Undo on Completion

tidy-dev committed:
Make reused method

steveward committed:
Make the linter happy

tidy-dev committed:
Typing by HTMLAttributes

Co-Authored-By: Sergio Padrino <1083228+sergiou87@users.noreply.github.com>

dependabot[bot] committed:
Bump peter-evans/create-pull-request from 4.2.3 to 4.2.4

Bumps [peter-evans/create-pull-request](https://github.com/peter-evans/create-pull-request) from 4.2.3 to 4.2.4.
- [Release notes](https://github.com/peter-evans/create-pull-request/releases)
- [Commits](https://github.com/peter-evans/create-pull-request/compare/v4.2.3...v4.2.4)

---
updated-dependencies:
- dependency-name: peter-evans/create-pull-request
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

tidy-dev committed:
More deterministic approach - think my brain wasn't working Friday

tidy-dev committed:
For first time undo displays

tidy-dev committed:
Match button text  - condense repeated code

tidy-dev committed:
Tidying..

tidy-dev committed:
Tidying.. - make sure amending gets right message

tidy-dev committed:
Tidying

tidy-dev committed:
Merge pull request #16346 from desktop/dependabot/github_actions/peter-evans/create-pull-request-4.2.4

Bump peter-evans/create-pull-request from 4.2.3 to 4.2.4

sergiou87 committed:
Merge pull request #16334 from desktop/pr-comment-notification-metrics

Track metrics for PR comment notifications

sergiou87 committed:
Merge pull request #16339 from desktop/fix-rstudio-macos

Add new RStudio bundle identifier

sergiou87 committed:
Replace dialog `header` with `div`

sergiou87 committed:
Merge pull request #16350 from desktop/dialog-header-is-div

Replace dialog `header` with `div`

tidy-dev committed:
Use HTMLElement

tidy-dev committed:
Revert undo focus changes

tidy-dev committed:
Announce just committed and delay removal of loading message

tidy-dev committed:
Use one message

sergiou87 committed:
Focus Changes/History view when accessed via shortcut

sergiou87 committed:
Use role=combobox for co-authors input

sergiou87 committed:
Refactor co-author render and allow tabbing into the list

sergiou87 committed:
Fix focusing added authors with (shift+)tab

niik committed:
Update app/src/lib/api.ts

Co-authored-by: Sergio Padrino <sergio.padrino@gmail.com>

tidy-dev committed:
Update commit-message.tsx

tidy-dev committed:
Merge pull request #16315 from desktop/windows-app-menu-screen-reader-fixes

Windows app menu screen reader fixes

tidy-dev committed:
Release 3.2.1-beta2

sergiou87 committed:
Merge pull request #16360 from desktop/focus-history-and-changes

tidy-dev committed:
Don't need mostRecentLocalCommit anymore

tidy-dev committed:
Delay if time > 0

tidy-dev committed:
Merge branch 'development' into releases/3.2.1-beta2

tidy-dev committed:
Add focus history/changes list fix

tidy-dev committed:
Merge pull request #16340 from desktop/annouce-commit-button-status

Announce Committing Status

tidy-dev committed:
Update changelog.json

Co-authored-by: Sergio Padrino <sergio.padrino@gmail.com>

tidy-dev committed:
Update changelog.json

Co-authored-by: Sergio Padrino <sergio.padrino@gmail.com>

sergiou87 committed:
Add remove button to author handles

dereklowlind committed:
update bitbucket.md integration docs

Bitbucket changed their UI so I refreshed this doc to reflect it

sergiou87 committed:
Update changelog.json

tidy-dev committed:
Use overflow-y auto not scroll

tidy-dev committed:
Merge pull request #16367 from desktop/windows-menu-scroll-overflow

Windows Menus: Use overflow-y auto not scroll

niik committed:
Lint

tidy-dev committed:
Merge pull request #16362 from desktop/releases/3.2.1-beta2

Release 3.2.1-beta2

niik committed:
Merge pull request #16338 from JosephTLyons/add-Zed-as-an-external-editor-option

Add Zed as an external editor option

tidy-dev committed:
Remove delay

sergiou87 committed:
Add warning for commits with unknown authors

sergiou87 committed:
Add styling for Unknown Authors dialog

sergiou87 committed:
Remove unnecessary CSS class

tidy-dev committed:
State not updated when actions are updated

tidy-dev committed:
Needs something new in the aria-live to ensure it is announced

tidy-dev committed:
Update to use resizeActiveResizable

tidy-dev committed:
Add resizable event listeners to resizable component

tidy-dev committed:
Remove unused methods from last approach

tidy-dev committed:
Need to be () =>

tidy-dev committed:
Typo

tidy-dev committed:
5px better than one

tidy-dev committed:
More concise words?

tidy-dev committed:
Correct word with action

tidy-dev committed:
Update menu state when resizable focus changes

tidy-dev committed:
Add broken logic to enable/disable menu increase when not applicable

tidy-dev committed:
Removed unused method

tidy-dev committed:
More removal

sergiou87 committed:
Merge pull request #16366 from dereklowlind/development

update bitbucket.md integration docs

sergiou87 committed:
Fix CSS of PR preview dialog

sergiou87 committed:
Merge pull request #16395 from desktop/preview-pr-header-css

Fix CSS of PR preview dialog

tidy-dev committed:
Merge pull request #16371 from desktop/Remove-unnecesssary-delay

Remove unnecessary delay in announcing committing status

sergiou87 committed:
Configure added co-authors as a listbox

tidy-dev committed:
Add those Ids - different accelerator to not mingle with zoom functionality

tidy-dev committed:
Use tracking of a focused elements because windows menus are elements

sergiou87 committed:
Add support private repo images

sergiou87 committed:
Add "Co-authors" label to input and list of coauthors

tidy-dev committed:
Merge branch 'development' into resize-pane-with-keyboard-shorcut

niik committed:
Merge pull request #16286 from desktop/stream-available-repos

Stream repositories available to the user from the API

tidy-dev committed:
Move misattributed popover into view when screen is small and  zoomed

tidy-dev committed:
Use button not anchor for banner close button

tidy-dev committed:
Remove other linters

tidy-dev committed:
make some div css apply to button

tidy-dev committed:
Make expander a button

Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum."

tidy-dev committed:
Override user agent button styles

tidy-dev committed:
Remove aria-selected attribute and linter

sergiou87 committed:
Use aria-labelledby for co-authors label references

sergiou87 committed:
Add auth token for github.com private asset requests

sergiou87 committed:
Rename filter to reflect its new behavior

sergiou87 committed:
Use nbsp to force screen readers to read number of suggestions when needed

tidy-dev committed:
Merge pull request #16408 from desktop/some-linters

Banner Close Button: Removing easy a11y linters

tidy-dev committed:
Merge pull request #16409 from desktop/expander-button-linters

Commit Detail Expander button linter cleanup

tidy-dev committed:
Merge pull request #16410 from desktop/stashed-changes-remove-aria-selector-linter

Stashed Changes Button: Remove aria-selected attribute and linter

sergiou87 committed:
Merge pull request #16335 from desktop/co-author-accessibility-fixes

Co-author revamped: less CodeMirror, more accessibility

niik committed:
Avoid retaining DOM elements

Keep track of just the state we're interested in

niik committed:
Announce the included state of changed file

tidy-dev committed:
Remove log

tidy-dev committed:
Use 8 and 9 as keyboard shortcut

Co-Authored-By: Markus Olsson <634063+niik@users.noreply.github.com>

niik committed:
Allow tabbing to empty lists

niik committed:
Announce empty lists

tidy-dev committed:
Merge pull request #16407 from desktop/misattribued-commit-popover-zoomed-

Move misattributed popover into view when screen is small and  zoomed

tidy-dev committed:
Merge pull request #16332 from desktop/resize-pane-with-keyboard-shorcut

Resize sidebar with keyboard shortcut

tidy-dev committed:
Condense commit description textarea when zoomed

tidy-dev committed:
Make changes list scroll when zoomed

tidy-dev committed:
group non-list divs

tidy-dev committed:
Use value not clamp

niik committed:
Merge pull request #16412 from desktop/private-repo-assets

Prepare for asset auth changes

KendallDoesCoding committed:
Merge branch 'desktop:development' into patch-2

niik committed:
Changelog and version bump

tidy-dev committed:
Merge pull request #16293 from KendallDoesCoding/patch-2

docs: small edits in grammar in Building Desktop section.

tidy-dev committed:
Use toggle tip component based on tooltip

niik committed:
Merge pull request #16445 from desktop/releases/3.2.1-beta3

Release 3.2.1-beta3

niik committed:
Don't introduce a11y props abstraction at this time

niik committed:
Camel case aria label prop

niik committed:
Changelog and version bump

tidy-dev committed:
Remove pr comments release notes

tidy-dev committed:
Remove non applicable release note

tidy-dev committed:
Remove duplicate release note

tidy-dev committed:
Merge pull request #16452 from desktop/releases/3.2.1

Release 3.2.1

tidy-dev committed:
Add aria described by to changed files checkbox

tidy-dev committed:
words

tidy-dev committed:
Update app/src/ui/lib/checkbox.tsx

Co-authored-by: Markus Olsson <j.markus.olsson@gmail.com>

tidy-dev committed:
Update app/src/ui/lib/checkbox.tsx

Co-authored-by: Markus Olsson <j.markus.olsson@gmail.com>

niik committed:
Merge pull request #16420 from desktop/listitem-aria-select

Announce file selected (as in included in commit) state

tidy-dev committed:
Contrain value respects min for max.

tidy-dev committed:
Use clamp again for sidebar

tidy-dev committed:
Other ariaDescribedBy spot

tidy-dev committed:
Merge branch 'development' into change-list-scrollable-zoomed

tidy-dev committed:
toolTipVisible should be state, subscribe to button ref in constructor

tidy-dev committed:
Add/remove tooltip visibility listeners

tidy-dev committed:
Merge branch 'development' into character-length-warning-keyboard-accessible

tidy-dev committed:
Merge pull request #16370 from desktop/fix-pr-preview--sometimes-erroneously-enabled

Suggested Next Action Dropdown State not updated when actions are updated

tidy-dev committed:
Merge pull request #16457 from desktop/add-description-for-changes-list-header-checkbox

Add description for changes list header checkbox

tidy-dev committed:
Make changes list panel and history log list be the tab associated panels

tidy-dev committed:
Focus first focusable element of changes panel

tidy-dev committed:
Merge branch 'development' into change-list-scrollable-zoomed

sweezyio committed:
Add icons for tabs in Repository settings dialog

tidy-dev committed:
Merge pull request #16425 from desktop/change-list-scrollable-zoomed

Change list scrollable when zoomed

tidy-dev committed:
Keep a button ref instance to remove listeners

tidy-dev committed:
Merge branch 'development' into character-length-warning-keyboard-accessible

tidy-dev committed:
Merge branch 'development' into remove-avatar-tooltip

tidy-dev committed:
Merge branch 'character-length-warning-keyboard-accessible' into remove-avatar-tooltip

tidy-dev committed:
Revert title removal stuff

tidy-dev committed:
Use toggle tip for commit message avatar

tidy-dev committed:
No tooltip when misattributed warning

tidy-dev committed:
Put username in warning

tidy-dev committed:
Add role of `alertdialog` implementation

tidy-dev committed:
Type to require ariaDescribedBy for alertdialog roles

Other IDialogProp things

shiftkey committed:
pull in additional stylings to allow scrollbar styling on Linux to match Windows

adapt scrollbar styling on linux for dark mode

shiftkey committed:
opt-out of rendering HTML menu on Linux

shiftkey committed:
enable auto-updating of theme and following system theme on Linux

tidy-dev committed:
Open the changed file header tool tip on checkbox focus

tidy-dev committed:
Add errors when too many popups

tidy-dev committed:
Update popup-manager.ts

dependabot[bot] committed:
Bump xml2js from 0.4.19 to 0.5.0

Bumps [xml2js](https://github.com/Leonidas-from-XIV/node-xml2js) from 0.4.19 to 0.5.0.
- [Release notes](https://github.com/Leonidas-from-XIV/node-xml2js/releases)
- [Commits](https://github.com/Leonidas-from-XIV/node-xml2js/compare/0.4.19...0.5.0)

---
updated-dependencies:
- dependency-name: xml2js
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>

GitMensch committed:
fix VSCodium name

as it never had a "Visual Studio" prefix, which is a registered name

niik committed:
Merge pull request #16511 from GitMensch/patch-2

fix VSCodium name

Commit:
Add VimR to editors/darwin.ts

Commit:
Add VimR to editor-integration.md MacOS list

niik committed:
Merge pull request #16465 from sweezyio/development

Add icons for tabs in Repository settings dialog

niik committed:
Merge pull request #16503 from desktop/dependabot/npm_and_yarn/xml2js-0.5.0

Bump xml2js from 0.4.19 to 0.5.0

shiftkey committed:
Update app/styles/themes/_dark.scss

Co-authored-by: Markus Olsson <j.markus.olsson@gmail.com>

niik committed:
Merge pull request #16515 from Elmar-Wiese/development

Add support for VimR code editor on macOS.

tidy-dev committed:
Merge pull request #16487 from desktop/open-changed-list-heading-tooltip-on-checkbox-focus

Open the changed file header tool tip on checkbox focus

tidy-dev committed:
Merge pull request #16499 from desktop/too-many-popups

Too many popups

tidy-dev committed:
Merge branch 'development' into undo-confirmation-alert-dialog

tidy-dev committed:
Cleaner interface union

tidy-dev committed:
Merge branch 'development' into character-length-warning-keyboard-accessible

tidy-dev committed:
Merge branch 'development' into changes-history-tabs

tidy-dev committed:
Merge pull request #16463 from desktop/changes-history-tabs

Changes and History tab panels are syntaxed as such for screen readers

tidy-dev committed:
Merge pull request #16449 from desktop/character-length-warning-keyboard-accessible

Use toggle tip component for commit summary length hint

tidy-dev committed:
Merge pull request #16272 from desktop/remove-avatar-tooltip

Tooltip for Committing Message Avatar is a toggle tip

niik committed:
Merge pull request #16484 from shiftkey/allow-for-dark-styling

tidy-dev committed:
Merge pull request #16472 from desktop/undo-confirmation-alert-dialog

Undo Confirmation Modal is an Alert Dialog

niik committed:
Merge pull request #14176 from desktop/shared-formatter-caching

Shared formatter caching

niik committed:
Merge branch 'locale-aware-formatting' into region-aware-number-formatting

niik committed:
Merge branch 'development' into region-aware-number-formatting

niik committed:
Merge fail
